pkgs/top-level/all-packages.nix

```diff
@@ -198,32 +198,28 @@ with pkgs;
     baseUrl = "https://${githubBase}/${owner}/${repo}";
     passthruAttrs = removeAttrs args [ "owner" "repo" "rev" "fetchSubmodules" "private" "githubBase" "varPrefix" ];
     varBase = "NIX${if varPrefix == null then "" else "_${varPrefix}"}_GITHUB_PRIVATE_";
-  in (if fetchSubmodules then
-    fetchgit ({
-      inherit name rev fetchSubmodules;
-      url = "${baseUrl}.git";
-    } // passthruAttrs)
-  else
     # We prefer fetchzip in cases we don't need submodules as the hash
     # is more stable in that case.
-    fetchzip ({
-      inherit name;
-      url = "${baseUrl}/archive/${rev}.tar.gz";
-    } // lib.optionalAttrs private {
-      netrcPhase = ''
-        if [ -z "''$${varBase}USERNAME" -o -z "''$${varBase}PASSWORD" ]; then
-          echo "Error: Private fetchFromGitHub requires the nix building process (nix-daemon in multi user mode) to have the ${varBase}USERNAME and ${varBase}PASSWORD env vars set." >&2
-          exit 1
-        fi
-        cat > netrc <<EOF
-        machine ${githubBase}
-                login ''$${varBase}USERNAME
-                password ''$${varBase}PASSWORD
-        EOF
-      '';
-      netrcImpureEnvVars = [ "${varBase}USERNAME" "${varBase}PASSWORD" ];
-    } // passthruAttrs) // { inherit rev; })
-  // { meta.homepage = "${baseUrl}/"; };
+    fetcher = if fetchSubmodules then fetchgit else fetchzip;
+    fetcherArgs = (if fetchSubmodules
+      then { inherit rev fetchSubmodules; url = "${baseUrl}.git"; }
+      else ({ url = "${baseUrl}/archive/${rev}.tar.gz"; } //
+        lib.optionalAttrs private {
+          netrcPhase = ''
+            if [ -z "''$${varBase}USERNAME" -o -z "''$${varBase}PASSWORD" ]; then
+              echo "Error: Private fetchFromGitHub requires the nix building process (nix-daemon in multi user mode) to have the ${varBase}USERNAME and ${varBase}PASSWORD env vars set." >&2
+              exit 1
+            fi
+            cat > netrc <<EOF
+            machine ${githubBase}
+                    login ''$${varBase}USERNAME
+                    password ''$${varBase}PASSWORD
+            EOF
+          '';
+          netrcImpureEnvVars = [ "${varBase}USERNAME" "${varBase}PASSWORD" ];
+        }))
+      // passthruAttrs // { inherit name; };
+  in fetcher fetcherArgs // { meta.homepage = "${baseUrl}/"; inherit rev; };
```

Comment:
I like this, thank you! (Although extracting the private section into `privateAttrs` will make it even better.)
I just noticed that GitHub does not add the final slash when you navigate to a project page, so we shouldn't too.

Reply:
Good points, thanks! I've changed it accordingly.